### PR TITLE
os/ObjectStore: properly clear object map when replaying OP_REMOVE

### PR DESCRIPTION
**Pull request by ukernel**
To remove an object, filestore needs to unlink corresponding object
file from filesystem and removes corresponding object keys from
DBObjectMap. When replaying OP_REMOVE operation, it's possible the
operation has completed partially, object file has been deleted, but
object keys in DBObjectMap hasn't.

The fix is force clear object keys if object file does not exists

Fixes: http://tracker.ceph.com/issues/17177
Signed-off-by: Yan, Zheng <zyan@redhat.com>

### CODE REVIEW

File: src/os/filestore/FileStore.cc [hunk 1]
```diff
@@ -472,13 +472,7 @@ int FileStore::lfn_unlink(const coll_t& cid, const ghobject_t& o,
...
 
     if (!force_clear_omap) {
-      if (hardlink == 0) {
-          if (!m_disable_wbthrottle) {
-	    wbthrottle.clear_object(o); // should be only non-cache ref
-          }
-	  fdcache.clear(o);
-	  return 0;
-      } else if (hardlink == 1) {
+      if (hardlink == 0 || hardlink == 1) {
```
Comment (tchaikov) [inline src/os/filestore/FileStore.cc]: we always clear omap first, and **then** unlink the object file. so, if the `hardlink` is  1 here, we should have cleared the omap already, am i right?

**Comment by ukernel** [comment]
ping @athanatos  @liewegas @tchaikov

File: src/os/filestore/FileStore.cc (hunk 1 above)
Comment (liewegas) [inline src/os/filestore/FileStore.cc]: So to clarify the situation is

 - clear omap
 - unlink
 - unlink persists to disk, but omap does not
 - <crash>
 - replay sees hardlink 0 and doesn't clear omap

?

Comment (ukernel) [inline src/os/filestore/FileStore.cc]: yes

**Comment by liewegas** [comment]
lgtm!

**Comment by athanatos** [comment]
lgtm! (Whoa, good catch!)

**Comment by tchaikov** [comment]
lgtm also.

**Comment by badone** [comment]
Thanks @ukernel Hopefully we can get this merged soon.

**Comment by yuriw** [comment]
http://pulpito.ceph.com/yuriw-2016-10-21_00:30:49-rados-wip-yuri-testing_2016_10_21-distro-basic-smithi/

http://pulpito.front.sepia.ceph.com/yuriw-2016-10-21_15:49:39-rados-wip-yuri-testing_2016_10_21-distro-basic-smithi/

**Pull request merged**